File: week-01/yc_2493_tower.java
import java.io.*;
import java.util.*;

public class Yc2493 {
    record Tower(int index, int height) {}

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        int[] towers = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++) {
            towers[i] = Integer.parseInt(st.nextToken());
        }

        Stack<Tower> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            int height = towers[i];

            while (!stack.isEmpty() && stack.peek().height() < height) {
                stack.pop();
            }

            if (!stack.isEmpty()) {
                sb.append(stack.peek().index()).append(" ");
            } else {
                sb.append("0 ");
            }

            stack.push(new Tower(i + 1, height));
        }

        System.out.println(sb.toString().trim());
    }
}

File: week-01/yc_2493_tower.cs
using System;
using System.Text;
using System.Collections.Generic;

record Tower(int Index, int Height);

public class Yc2493
{
    static void Main(string args[])
    {
        int n = int.Parse(Console.ReadLine());
        int[] towers = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
        StringBuilder sb = new StringBuilder();

        Stack<Tower> stack = new Stack<Tower>();

        for (int i = 0; i < n; i++)
        {
            int height = towers[i];

            while(stack.Count > 0 && stack.Peek().Height < height)
            {
                stack.Pop();
            }

            if (stack.Count > 0)
            {
                sb.Append(stack.Peek().Index).Append(" ");
            }
            else
            {
                sb.Append("0 ");
            }

            stack.Push(new Tower(i + 1, height));
        }
        Console.WriteLine(sb.ToString().Trim());
    }
}
